backend/routes/users.js

```diff
@@ -0,0 +1,77 @@
+const express = require('express');
+const bcrypt = require('bcrypt');
+const { PrismaClient } = require('@prisma/client');
+const prisma = new PrismaClient();
+
+const router = express.Router();
+
+router.post('/', async (req, res) => {
+    const { username, email, password } = req.body;
+
+    try {
+        const existingUser = await prisma.user.findFirst({
+            where: {
+                OR: [
+                    { username: username },
+                    { email: email }
+                ]
+            }
+        });
+
+        if (existingUser) {
+            return res.status(400).json({ error: 'Username or email already exists' });
+        }
+
+        const hashedPassword = await bcrypt.hash(password, 10);
+
+        const newUser = await prisma.user.create({
+            data: { username, email, password: hashedPassword },
+        });
+
+        req.session.user = newUser;
+
+        res.json({ user: newUser });
+    } catch (error) {
+        console.error('Error during user sign up:', error);
+        res.status(500).json({ error: 'Server error' });
+    }
+});
+
+router.post('/login', async (req, res) => {
+    const { username, password } = req.body;
+
+    try {
+        const user = await prisma.user.findFirst({
+            where: { username }
+        });
+
+        if (!user) {
+            return res.status(401).json({ error: 'Invalid username or password' });
+        }
+
+        const isValidPassword = await bcrypt.compare(password, user.password);
+
+        if (!isValidPassword) {
+            return res.status(401).json({ error: 'Invalid username or password' });
+        }
```

Comment:
Not sure if you did this intentionally, but it's a *good* thing that you are returning the same error in both of these cases. It is very common to return the exact same error whether the user does not exist or the password does not match, since that prevents attackers from having more information if they are brute-forcing many logins